File: core/tokens.py
"""64-token core of LOVA (Stage 1 MVP).

Each token is a single byte in [0x00, 0x3F]. Tokens are grouped into 8
families of 8. The full table is declared here; most are placeholders
in Milestone 1 — only the operators needed for the first end-to-end
experiment are implemented in ``core.runtime``. Touching any operator
unimplemented in the runtime raises ``NotImplementedError`` with a
pointer to the relevant family; this is intentional (Axiom 8 —
small core, every byte semantic).

See ``../spec/tokens.md`` for the human reference (DRAFT).
See ``../spec/axioms.md`` for the ten design invariants.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, List, Tuple, Union

# --- token table -------------------------------------------------------------

# Family 0x00-0x07  Structural
END             = 0x00   # list terminator for variadic ops
LIT_INT         = 0x01   # integer literal (length-prefixed big-endian follows)
PARTITION       = 0x02   # n → (a, b) with a+b=n, heat-directed split
MERGE           = 0x03   # a, b → a+b
HEAT_INC        = 0x04
HEAT_GET        = 0x05
INHERIT         = 0x06
IDENTITY        = 0x07

# Family 0x08-0x0F  Number theory
P               = 0x08   # partition number p(n)
TAU             = 0x09   # divisor count τ(n)
SIGMA           = 0x0A   # divisor sum σ(n)
PHI3            = 0x0B   # mock theta φ₃ (placeholder)
PSI7            = 0x0C   # mock theta ψ₇ (placeholder)
ETA             = 0x0D   # Dedekind η (placeholder)
GCD             = 0x0E
MOBIUS          = 0x0F

# Family 0x10-0x17  Conservation
BUDGET          = 0x10   # (budget k body) — body may consume ≤ k units
CONSERVE        = 0x11   # (conserve invariant body) — verify invariant holds
DELTA_CHECK     = 0x12
RESPAWN         = 0x13
BUDGET_REMAINING = 0x14
SUM_INVARIANT   = 0x15
PRESERVE        = 0x16
VIOLATE         = 0x17   # (violate)  — synthetic "break conservation" op, testing only

# Family 0x18-0x1F  Surprise / watch
SURPRISE        = 0x18   # (surprise predicted actual) — returns |p-a|, emits trace
WATCH           = 0x19
WHEN_ANOMALY    = 0x1A
THRESHOLD       = 0x1B
PREDICT         = 0x1C
TRACE_SURPRISE  = 0x1D
NORMAL_RANGE    = 0x1E
DEVIATION       = 0x1F

# Family 0x20-0x27  Evolution
DEFPOP          = 0x20
VARIANT         = 0x21
EVOLVE          = 0x22
SELECT          = 0x23
MUTATE          = 0x24
CLONE           = 0x25
FITNESS         = 0x26
RETIRE          = 0x27

# Family 0x28-0x2F  Composition
SEQ             = 0x28   # variadic sequential composition (terminated by END)
PAR             = 0x29
IF_SURPRISE     = 0x2A
LOOP_UNTIL      = 0x2B
LAMBDA          = 0x2C   # (lambda (params...) body) — params is a LIT_INT symbol list
APPLY           = 0x2D
LET             = 0x2E   # (let name value body)
REF             = 0x2F   # (ref name) — dereference bound name

# Family 0x30-0x37  Effects / IO (all placeholders in MVP)
EXTERNAL_BOUNDARY = 0x30
NET_SEND        = 0x31
NET_RECV        = 0x32
FS_READ         = 0x33
FS_WRITE        = 0x34
STDOUT          = 0x35
STDIN           = 0x36
CLOCK           = 0x37

# Family 0x38-0x3F  Meta / lineage
LINEAGE_QUERY   = 0x38
WHY             = 0x39
TRACE           = 0x3A
EXPLAIN         = 0x3B
HASH            = 0x3C
UID             = 0x3D
ANCESTOR_OF     = 0x3E
GENERATION      = 0x3F


# --- signatures (arity + semantic family for the decoder) -------------------

# arity: an int (fixed arity) or the string "variadic" (terminated by END).
# LIT_INT has arity 0 but a special length-prefixed payload.
SIGNATURES = {
    # Structural
    END:          {"name": "end",           "arity": 0, "family": "struct"},
    LIT_INT:      {"name": "lit",           "arity": 0, "family": "struct",
                   "payload": "varint"},
    PARTITION:    {"name": "partition",     "arity": 1, "family": "struct"},
    MERGE:        {"name": "merge",         "arity": 2, "family": "struct"},
    HEAT_INC:     {"name": "heat-inc",      "arity": 1, "family": "struct"},
    HEAT_GET:     {"name": "heat-get",      "arity": 1, "family": "struct"},
    INHERIT:      {"name": "inherit",       "arity": 2, "family": "struct"},
    IDENTITY:     {"name": "identity",      "arity": 1, "family": "struct"},
    # Number theory
    P:            {"name": "p",             "arity": 1, "family": "nt"},
    TAU:          {"name": "tau",           "arity": 1, "family": "nt"},
    SIGMA:        {"name": "sigma",         "arity": 1, "family": "nt"},
    PHI3:         {"name": "phi3",          "arity": 2, "family": "nt"},
    PSI7:         {"name": "psi7",          "arity": 2, "family": "nt"},
    ETA:          {"name": "eta",           "arity": 1, "family": "nt"},
    GCD:          {"name": "gcd",           "arity": 2, "family": "nt"},
    MOBIUS:       {"name": "mobius",        "arity": 1, "family": "nt"},
    # Conservation
    BUDGET:       {"name": "budget",        "arity": 2, "family": "cons"},
    CONSERVE:     {"name": "conserve",      "arity": 2, "family": "cons"},
    DELTA_CHECK:  {"name": "delta-check",   "arity": 1, "family": "cons"},
    RESPAWN:      {"name": "respawn",       "arity": 1, "family": "cons"},
    BUDGET_REMAINING: {"name": "budget-remaining", "arity": 0, "family": "cons"},
    SUM_INVARIANT: {"name": "sum-invariant", "arity": 0, "family": "cons"},
    PRESERVE:     {"name": "preserve",      "arity": 2, "family": "cons"},
    VIOLATE:      {"name": "violate",       "arity": 1, "family": "cons"},
    # Surprise
    SURPRISE:     {"name": "surprise",      "arity": 2, "family": "surp"},
    WATCH:        {"name": "watch",         "arity": 1, "family": "surp"},
    WHEN_ANOMALY: {"name": "when-anomaly",  "arity": 2, "family": "surp"},
    THRESHOLD:    {"name": "threshold",     "arity": 1, "family": "surp"},
    PREDICT:      {"name": "predict",       "arity": 1, "family": "surp"},
    TRACE_SURPRISE: {"name": "trace-surprise", "arity": 1, "family": "surp"},
    NORMAL_RANGE: {"name": "normal-range",  "arity": 2, "family": "surp"},
    DEVIATION:    {"name": "deviation",     "arity": 2, "family": "surp"},
    # Evolution
    DEFPOP:       {"name": "defpop",        "arity": "variadic", "family": "evo"},
    VARIANT:      {"name": "variant",       "arity": 2, "family": "evo"},
    EVOLVE:       {"name": "evolve",        "arity": 1, "family": "evo"},
    SELECT:       {"name": "select",        "arity": 2, "family": "evo"},
    MUTATE:       {"name": "mutate",        "arity": 2, "family": "evo"},
    CLONE:        {"name": "clone",         "arity": 1, "family": "evo"},
    FITNESS:      {"name": "fitness",       "arity": 1, "family": "evo"},
    RETIRE:       {"name": "retire",        "arity": 1, "family": "evo"},
    # Composition
    SEQ:          {"name": "seq",           "arity": "variadic", "family": "comp"},
    PAR:          {"name": "par",           "arity": "variadic", "family": "comp"},
    IF_SURPRISE:  {"name": "if-surprise",   "arity": 3, "family": "comp"},
    LOOP_UNTIL:   {"name": "loop-until",    "arity": 2, "family": "comp"},
    LAMBDA:       {"name": "lambda",        "arity": 2, "family": "comp"},
    APPLY:        {"name": "apply",         "arity": "variadic", "family": "comp"},
    LET:          {"name": "let",           "arity": 3, "family": "comp"},
    REF:          {"name": "ref",           "arity": 1, "family": "comp"},
    # IO (placeholders)
    EXTERNAL_BOUNDARY: {"name": "external-boundary", "arity": 2, "family": "io"},
    NET_SEND:     {"name": "net-send",      "arity": 1, "family": "io"},
    NET_RECV:     {"name": "net-recv",      "arity": 0, "family": "io"},
    FS_READ:      {"name": "fs-read",       "arity": 1, "family": "io"},
    FS_WRITE:     {"name": "fs-write",      "arity": 2, "family": "io"},
    STDOUT:       {"name": "stdout",        "arity": 1, "family": "io"},
    STDIN:        {"name": "stdin",         "arity": 0, "family": "io"},
    CLOCK:        {"name": "clock",         "arity": 0, "family": "io"},
    # Meta
    LINEAGE_QUERY: {"name": "lineage-query", "arity": 1, "family": "meta"},
    WHY:          {"name": "why",           "arity": 1, "family": "meta"},
    TRACE:        {"name": "trace",         "arity": 1, "family": "meta"},
    EXPLAIN:      {"name": "explain",       "arity": 1, "family": "meta"},
    HASH:         {"name": "hash",          "arity": 1, "family": "meta"},
    UID:          {"name": "uid",           "arity": 0, "family": "meta"},
    ANCESTOR_OF:  {"name": "ancestor-of",   "arity": 2, "family": "meta"},
    GENERATION:   {"name": "generation",    "arity": 1, "family": "meta"},
}

assert len(SIGNATURES) == 64, f"Token table must have exactly 64 entries, found {len(SIGNATURES)}"


# --- typed-slot extensions (Milestone 2) ------------------------------------
#
# Only operators implemented by the Milestone 1 runtime receive type
# annotations.  Reserved operators have no ``out_type``; the generator
# treats them as unreachable and never emits them.  As milestones land,
# operators get type info here and become generation-reachable.

from core.types import INT, LITERAL_INT  # noqa: E402

_TYPE_INFO = {
    # Literals
    LIT_INT:        {"in_types": [], "out_type": LITERAL_INT},
    # Structural (M1 subset)
    PARTITION:      {"in_types": [INT], "out_type": INT},
    MERGE:          {"in_types": [INT, INT], "out_type": INT},
    IDENTITY:       {"in_types": [INT], "out_type": INT},
    # Number theory (M1 subset)
    P:              {"in_types": [INT], "out_type": INT},
    TAU:            {"in_types": [INT], "out_type": INT},
    SIGMA:          {"in_types": [INT], "out_type": INT},
    GCD:            {"in_types": [INT, INT], "out_type": INT},
    MOBIUS:         {"in_types": [INT], "out_type": INT},
    # Conservation (M1 subset)
    BUDGET:         {"in_types": [LITERAL_INT, INT], "out_type": INT},
    CONSERVE:       {"in_types": [INT, INT], "out_type": INT},
    VIOLATE:        {"in_types": [INT], "out_type": INT},
    # Surprise (M1 subset)
    SURPRISE:       {"in_types": [INT, INT], "out_type": INT},
    TRACE_SURPRISE: {"in_types": [INT], "out_type": INT},
    # Composition (M1 subset)
    SEQ:            {"in_types": None, "variadic_type": INT, "out_type": INT},
    LET:            {"in_types": [LITERAL_INT, INT, INT], "out_type": INT},
    REF:            {"in_types": [LITERAL_INT], "out_type": INT},
    IF_SURPRISE:    {"in_types": [INT, INT, INT], "out_type": INT},
    # END is a structural sentinel — no out_type; the generator
    # handles it specially as a variadic terminator.
}

for _tok, _extra in _TYPE_INFO.items():
    SIGNATURES[_tok].update(_extra)

# Tokens currently reachable by type-directed generation.
TYPED_TOKENS = frozenset(_TYPE_INFO.keys())

# Symbol (interned name) → token byte.  Used by the surface parser.
NAME_TO_TOKEN = {sig["name"]: tok for tok, sig in SIGNATURES.items()}

# Unicode / ASCII aliases for Stage-1 readability.  These are surface-only;
# the substrate only knows NAME_TO_TOKEN.
ALIASES = {
    "⊕": "merge",
    "⊖": "partition",
    "τ": "tau",
    "σ": "sigma",
    "φ₃": "phi3",
    "ψ₇": "psi7",
    "η": "eta",
    "μ": "mobius",
    "λ": "lambda",
    "?":  "surprise",
}


# --- AST ---------------------------------------------------------------------

@dataclass
class Node:
    """Decoded program node — token + children (or an integer payload for LIT_INT)."""
    op: int
    args: List[Union["Node", int]]

    def __repr__(self) -> str:
        if self.op == LIT_INT:
            return f"Lit({self.args[0]})"
        name = SIGNATURES[self.op]["name"]
        return f"({name} {' '.join(repr(a) for a in self.args)})"


# --- encoding ----------------------------------------------------------------

def encode(node: Node) -> bytes:
    """Serialise a Node tree to a byte stream.

    LIT_INT uses a 1-byte length prefix (little surprise — Milestone 1
    picks a simple, obviously-lossless encoding over a space-optimal
    varint).  Max literal size is 255 bytes.  Values are signed, big-endian,
    two's complement (via Python ``int.to_bytes(..., signed=True)``).
    """
    buf = bytearray()
    _encode_into(node, buf)
    return bytes(buf)


def _encode_into(node: Node, buf: bytearray) -> None:
    buf.append(node.op)
    if node.op == LIT_INT:
        val: int = node.args[0]
        # How many bytes needed?  ``bit_length`` + sign bit.
        bits = val.bit_length() + 1
        n_bytes = max(1, (bits + 7) // 8)
        if n_bytes > 255:
            raise ValueError(f"LIT_INT too large: {val}")
        buf.append(n_bytes)
        buf.extend(val.to_bytes(n_bytes, "big", signed=True))
        return
    sig = SIGNATURES[node.op]
    if sig["arity"] == "variadic":
        for child in node.args:
            _encode_into(child, buf)
        buf.append(END)
    else:
        expected = sig["arity"]
        if len(node.args) != expected:
            raise ValueError(
                f"token {sig['name']} expects {expected} args, got {len(node.args)}"
            )
        for child in node.args:
            _encode_into(child, buf)


# --- decoding ----------------------------------------------------------------

def decode(data: bytes) -> Node:
    """Decode a byte stream into a Node.  The stream is expected to contain
    exactly one top-level expression."""
    node, pos = _decode_one(data, 0)
    if pos != len(data):
        trailing = data[pos:].hex()
        raise ValueError(f"trailing bytes after decode: {trailing}")
    return node


def _decode_one(data: bytes, pos: int) -> Tuple[Node, int]:
    if pos >= len(data):
        raise ValueError("unexpected end of stream")
    op = data[pos]
    if op not in SIGNATURES:
        raise ValueError(f"unknown token 0x{op:02X} at position {pos}")
    if op == LIT_INT:
        if pos + 1 >= len(data):
            raise ValueError("LIT_INT: missing length byte")
        length = data[pos + 1]
        if pos + 2 + length > len(data):
            raise ValueError("LIT_INT: truncated payload")
        val = int.from_bytes(data[pos + 2 : pos + 2 + length], "big", signed=True)
        return Node(op=LIT_INT, args=[val]), pos + 2 + length
    sig = SIGNATURES[op]
    pos += 1
    children: List[Any] = []
    if sig["arity"] == "variadic":
        while pos < len(data) and data[pos] != END:
            child, pos = _decode_one(data, pos)
            children.append(child)
        if pos >= len(data):
            raise ValueError(f"{sig['name']}: missing END terminator")
        pos += 1  # consume END
    else:
        for _ in range(sig["arity"]):
            child, pos = _decode_one(data, pos)
            children.append(child)
    return Node(op=op, args=children), pos


# --- convenience constructors ------------------------------------------------

def Lit(value: int) -> Node:
    return Node(op=LIT_INT, args=[value])


def Call(op_name: str, *args: Union[Node, int]) -> Node:
    """Build a Node by operator name (resolves aliases)."""
    actual_name = ALIASES.get(op_name, op_name)
    tok = NAME_TO_TOKEN[actual_name]
    children = [Lit(a) if isinstance(a, int) else a for a in args]
    return Node(op=tok, args=children)


# --- self-test ---------------------------------------------------------------

def _self_test() -> None:
    """Basic round-trip check.  Run via ``python -m core.tokens``."""
    tree = Call("p", 12)
    data = encode(tree)
    back = decode(data)
    assert back == tree, f"round-trip mismatch:\n  in:  {tree}\n  out: {back}"
    # nested
    tree2 = Call("merge", Call("p", 3), Call("tau", 12))
    data2 = encode(tree2)
    back2 = decode(data2)
    assert back2 == tree2
    # variadic SEQ
    tree3 = Call("seq", Call("p", 3), Call("p", 4), Call("p", 5))
    data3 = encode(tree3)
    back3 = decode(data3)
    assert back3 == tree3
    print("core.tokens self-test OK")
    print(f"  (p 12)                        -> {data.hex(' ')}  ({len(data)} bytes)")
    print(f"  (merge (p 3) (tau 12))        -> {data2.hex(' ')}  ({len(data2)} bytes)")
    print(f"  (seq (p 3) (p 4) (p 5))       -> {data3.hex(' ')}  ({len(data3)} bytes)")


if __name__ == "__main__":
    _self_test()

File: native/lova-rt/src/tokens.rs
//! The byte encoding (spec §1): opcodes, arities, and `decode`.

use crate::int::Int;
use std::cell::Cell;
use std::rc::Rc;

pub enum Arity {
    N(u8),
    Var,
}

pub struct Sig {
    pub name: &'static str,
    pub arity: Arity,
}

pub use crate::tokens_table::SIGS;

pub fn sig(op: u8) -> Option<&'static Sig> {
    SIGS.get(op as usize)
}

pub fn op_name(op: u8) -> &'static str {
    match sig(op) {
        Some(s) => s.name,
        None => "?",
    }
}

// --- opcodes ----------------------------------------------------------------

pub const END: u8 = 0x00;
pub const LIT_INT: u8 = 0x01;
pub const PARTITION: u8 = 0x02;
pub const MERGE: u8 = 0x03;
pub const CONS: u8 = 0x04;
pub const HEAD: u8 = 0x05;
pub const TAIL: u8 = 0x06;
pub const IDENTITY: u8 = 0x07;
pub const P: u8 = 0x08;
pub const TAU: u8 = 0x09;
pub const SIGMA: u8 = 0x0A;
pub const MUL: u8 = 0x0B;
pub const MOD: u8 = 0x0C;
pub const DIV: u8 = 0x0D;
pub const GCD: u8 = 0x0E;
pub const MOBIUS: u8 = 0x0F;
pub const BUDGET: u8 = 0x10;
pub const CONSERVE: u8 = 0x11;
pub const SIGNAL: u8 = 0x12;
pub const MAP_PUT: u8 = 0x13;
pub const MAP_GET: u8 = 0x14;
pub const NIL: u8 = 0x15;
pub const MAP_PAIRS: u8 = 0x16;
pub const VIOLATE: u8 = 0x17;
pub const SURPRISE: u8 = 0x18;
pub const IS_NIL: u8 = 0x19;
pub const WHEN_ANOMALY: u8 = 0x1A;
pub const THRESHOLD: u8 = 0x1B;
pub const EVAL: u8 = 0x1C;
pub const TRACE_SURPRISE: u8 = 0x1D;
pub const READ: u8 = 0x1E;
pub const DEVIATION: u8 = 0x1F;
pub const DEFPOP: u8 = 0x20;
pub const VARIANT: u8 = 0x21;
pub const EVOLVE: u8 = 0x22;
pub const SELECT: u8 = 0x23;
pub const MUTATE: u8 = 0x24;
pub const CLONE: u8 = 0x25;
pub const FITNESS: u8 = 0x26;
pub const RETIRE: u8 = 0x27;
pub const NET_SEND: u8 = 0x31;
pub const NET_RECV: u8 = 0x32;
pub const FS_READ: u8 = 0x33;
pub const FS_WRITE: u8 = 0x34;
pub const STDOUT: u8 = 0x35;
pub const STDIN: u8 = 0x36;
pub const CLOCK: u8 = 0x37;
pub const SEQ: u8 = 0x28;
pub const QUOTE: u8 = 0x29;
pub const IF_SURPRISE: u8 = 0x2A;
pub const LOOP_UNTIL: u8 = 0x2B;
pub const LAMBDA: u8 = 0x2C;
pub const APPLY: u8 = 0x2D;
pub const LET: u8 = 0x2E;
pub const REF: u8 = 0x2F;
pub const EXTERNAL_BOUNDARY: u8 = 0x30;
pub const LINEAGE_QUERY: u8 = 0x38;
pub const WHY: u8 = 0x39;
pub const TRACE: u8 = 0x3A;
pub const EXPLAIN: u8 = 0x3B;
pub const HASH: u8 = 0x3C;
pub const UID: u8 = 0x3D;
pub const ANCESTOR_OF: u8 = 0x3E;
pub const GENERATION: u8 = 0x3F;
pub const LIT_TEXT: u8 = 0x40;
pub const TEXT_LEN: u8 = 0x41;
pub const TEXT_CAT: u8 = 0x42;
pub const TEXT_SLICE: u8 = 0x43;
pub const TEXT_FIND: u8 = 0x44;
pub const TEXT_SPLIT: u8 = 0x45;
pub const TEXT_JOIN: u8 = 0x46;
pub const TEXT_CHARS: u8 = 0x47;
pub const TEXT_OF_CHARS: u8 = 0x48;
pub const TEXT_CMP: u8 = 0x49;
pub const TEXT_INT: u8 = 0x4A;
pub const INT_TEXT: u8 = 0x4B;
pub const IS_TEXT: u8 = 0x4C;
pub const TEXT_TRIM: u8 = 0x4D;
pub const TEXT_MATCH: u8 = 0x4E;
pub const TEXT_MATCH_ALL: u8 = 0x4F;
pub const LIST_MAP: u8 = 0x50;
pub const LIST_FILTER: u8 = 0x51;
pub const LIST_FOLD: u8 = 0x52;
pub const LIST_REVERSE: u8 = 0x53;
pub const LIST_RANGE: u8 = 0x54;
pub const LIST_ANY: u8 = 0x55;
pub const LIST_SORT_BY: u8 = 0x56;
pub const LIST_ZIP: u8 = 0x57;

// --- capabilities (spec §5.7) ----------------------------------------------

pub const CAP_FS_READ: u32 = 1;
pub const CAP_FS_WRITE: u32 = 2;
pub const CAP_CLOCK: u32 = 4;
pub const CAP_NET: u32 = 8;

pub fn capability_names(mask: u32) -> Vec<&'static str> {
    let mut out = Vec::new();
    if mask & CAP_FS_READ != 0 {
        out.push("fs-read");
    }
    if mask & CAP_FS_WRITE != 0 {
        out.push("fs-write");
    }
    if mask & CAP_CLOCK != 0 {
        out.push("clock");
    }
    if mask & CAP_NET != 0 {
        out.push("net");
    }
    out
}

/// A Python list of names, formatted as `str(list)` renders it.
pub fn py_name_list(names: &[&str]) -> String {
    let inner: Vec<String> = names.iter().map(|n| format!("'{}'", n)).collect();
    format!("[{}]", inner.join(", "))
}

// --- the node arena ---------------------------------------------------------

pub struct Node {
    pub op: u8,
    pub kids: Vec<u32>,
    pub ival: Option<Int>,
    pub sval: Option<Rc<String>>,
    /// The lineage uid, set by `register_root` / `_register_child`.
    /// Metadata on the node, never in the bytes (spec §5.5).
    pub uid: Cell<Option<u64>>,
}

pub fn node(op: u8, kids: Vec<u32>, ival: Option<Int>, sval: Option<Rc<String>>) -> Node {
    Node { op, kids, ival, sval, uid: Cell::new(None) }
}

pub struct Arena {
    pub nodes: Vec<Node>,
}

impl Arena {
    pub fn new() -> Arena {
        Arena { nodes: Vec::new() }
    }

    #[inline]
    pub fn get(&self, id: u32) -> &Node {
        &self.nodes[id as usize]
    }

    #[inline]
    pub fn op(&self, id: u32) -> u8 {
        self.nodes[id as usize].op
    }

    #[inline]
    pub fn kids(&self, id: u32) -> &[u32] {
        &self.nodes[id as usize].kids
    }

    pub fn push(&mut self, node: Node) -> u32 {
        self.nodes.push(node);
        (self.nodes.len() - 1) as u32
    }

    /// `_deep_copy_node`: a fresh tree, sharing a `LIT_TEXT`'s text.
    pub fn deep_copy(&mut self, id: u32) -> u32 {
        let (op, kids, ival, sval) = {
            let n = self.get(id);
            (n.op, n.kids.clone(), n.ival.clone(), n.sval.clone())
        };
        let new_kids: Vec<u32> = kids.iter().map(|k| self.deep_copy(*k)).collect();
        // `_deep_copy_node` clears the uid on the copy (quirk 29).
        self.push(node(op, new_kids, ival, sval))
    }
}

// --- decode (spec §1.1) -----------------------------------------------------

pub fn decode(arena: &mut Arena, data: &[u8]) -> Result<u32, String> {
    let (id, pos) = decode_one(arena, data, 0)?;
    if pos != data.len() {
        return Err(format!("trailing bytes after decode: {}", hex::encode(&data[pos..])));
    }
    Ok(id)
}

fn decode_one(arena: &mut Arena, data: &[u8], mut pos: usize) -> Result<(u32, usize), String> {
    if pos >= data.len() {
        return Err("unexpected end of stream".to_string());
    }
    let op = data[pos];
    let s = match sig(op) {
        Some(s) => s,
        None => return Err(format!("unknown token 0x{:02X} at position {}", op, pos)),
    };
    if op == LIT_INT {
        if pos + 1 >= data.len() {
            return Err("LIT_INT: missing length byte".to_string());
        }
        let length = data[pos + 1] as usize;
        if pos + 2 + length > data.len() {
            return Err("LIT_INT: truncated payload".to_string());
        }
        let val = Int::from_signed_bytes_be(&data[pos + 2..pos + 2 + length]);
        let id = arena.push(node(op, Vec::new(), Some(val), None));
        return Ok((id, pos + 2 + length));
    }
    if op == LIT_TEXT {
        if pos + 3 > data.len() {
            return Err("LIT_TEXT: missing length".to_string());
        }
        let length = ((data[pos + 1] as usize) << 8) | data[pos + 2] as usize;
        if pos + 3 + length > data.len() {
            return Err("LIT_TEXT: truncated payload".to_string());
        }
        let text = match std::str::from_utf8(&data[pos + 3..pos + 3 + length]) {
            Ok(t) => t.to_string(),
            Err(e) => return Err(format!("LIT_TEXT: invalid utf-8: {}", e)),
        };
        let id = arena.push(node(op, Vec::new(), None, Some(Rc::new(text))));
        return Ok((id, pos + 3 + length));
    }
    pos += 1;
    let mut kids: Vec<u32> = Vec::new();
    match s.arity {
        Arity::Var => {
            while pos < data.len() && data[pos] != END {
                let (child, next) = decode_one(arena, data, pos)?;
                kids.push(child);
                pos = next;
            }
            if pos >= data.len() {
                return Err(format!("{}: missing END terminator", s.name));
            }
            pos += 1;
        }
        Arity::N(n) => {
            for _ in 0..n {
                let (child, next) = decode_one(arena, data, pos)?;
                kids.push(child);
                pos = next;
            }
        }
    }
    let id = arena.push(node(op, kids, None, None));
    Ok((id, pos))
}

// --- encode (core/tokens.py) ------------------------------------------------

/// `encode(node)`: the byte stream `hash` turns into an integer.
pub fn encode(arena: &Arena, id: u32) -> Vec<u8> {
    let mut buf = Vec::new();
    encode_into(arena, id, &mut buf);
    buf
}

fn encode_into(arena: &Arena, id: u32, buf: &mut Vec<u8>) {
    let n = arena.get(id);
    buf.push(n.op);
    if n.op == LIT_TEXT {
        let data = n.sval.as_ref().unwrap().as_bytes();
        buf.push((data.len() >> 8) as u8);
        buf.push((data.len() & 0xFF) as u8);
        buf.extend_from_slice(data);
        return;
    }
    if n.op == LIT_INT {
        let value = n.ival.as_ref().unwrap();
        // `bit_length() + 1` for the sign, rounded up to whole bytes.
        let bits = value.bit_length() + 1;
        let n_bytes = std::cmp::max(1, ((bits + 7) / 8) as usize);
        let mut bytes = value.to_big().to_signed_bytes_be();
        if bytes.len() < n_bytes {
            let fill = if value.is_negative() { 0xFF } else { 0x00 };
            let mut padded = vec![fill; n_bytes - bytes.len()];
            padded.extend_from_slice(&bytes);
            bytes = padded;
        }
        buf.push(n_bytes as u8);
        buf.extend_from_slice(&bytes[bytes.len() - n_bytes..]);
        return;
    }
    for k in &n.kids {
        encode_into(arena, *k, buf);
    }
    if matches!(sig(n.op).map(|s| &s.arity), Some(Arity::Var)) {
        buf.push(END);
    }
}

/// `core.surface.pretty` -- the Stage-1 projection, which
/// `format_value` needs for a program value.
pub fn pretty(arena: &Arena, id: u32) -> String {
    let n = arena.get(id);
    if n.op == LIT_INT {
        return n.ival.as_ref().unwrap().to_string();
    }
    if n.op == LIT_TEXT {
        return crate::value::quote_text(n.sval.as_ref().unwrap());
    }
    let name = op_name(n.op);
    if n.kids.is_empty() {
        return format!("({})", name);
    }
    let kids: Vec<String> = n.kids.iter().map(|k| pretty(arena, *k)).collect();
    format!("({} {})", name, kids.join(" "))
}

/// `ordinals[node id]` = the node's 0-based place in the byte stream,
/// which is preorder with literals counted -- the order
/// `core.tokens.decode` meets them in.  Only the decoded tree is in the
/// table; a node made later (`quote`, `clone`, `mutate`, a probe) is
/// past its end and has no place in the program's text.
pub fn preorder_ordinals(arena: &Arena, root: u32) -> Vec<u32> {
    let mut out = vec![0u32; arena.nodes.len()];
    let mut next = 0u32;
    fill(arena, root, &mut next, &mut out);
    out
}

fn fill(arena: &Arena, id: u32, next: &mut u32, out: &mut Vec<u32>) {
    out[id as usize] = *next;
    *next += 1;
    for k in arena.kids(id) {
        fill(arena, *k, next, out);
    }
}

/// The set of operators a tree uses -- the scan D7 asks for before a run.
pub fn operators_used(arena: &Arena, id: u32, seen: &mut [bool; 0x58]) {
    let n = arena.get(id);
    seen[n.op as usize] = true;
    for k in &n.kids {
        operators_used(arena, *k, seen);
    }
}
